### PR TITLE
Refine vacations pipeline (Part 3. Final)

### PR DESCRIPTION
**Pull request by sponkrashin**
Issue #291

### CODE REVIEW

File: server/Arcadia.Assistant.CSP/CspCalendarEventsApprovalsChecker.cs [hunk 1]
```diff
@@ -7,26 +7,62 @@
...
 
-        protected override async Task<string> GetNextApprover(string employeeId, IEnumerable<string> existingApprovals)
+        protected override async Task<string> GetNextApprover(string employeeId, IEnumerable<string> existingApprovals, string eventType)
         {
             var employeesActor = Context.ActorSelection(EmployeesStorageActorPath);
             var departmentsActor = Context.ActorSelection(DepartmentsStorageActorPath);
 
             var allDepartments = await this.GetDepartments(departmentsActor);
-            var employeeDepartmentId = await this.GetEmployeeDepartmentId(employeesActor, employeeId);
+            var employeeMetadata = await this.GetEmployeeDepartmentId(employeesActor, employeeId);
```
Comment (santee) [inline server/Arcadia.Assistant.CSP/CspCalendarEventsApprovalsChecker.cs]: employeeMetadata = GetEmployeeDepartmentId? :)

File: server/Arcadia.Assistant.CSP/CspCalendarEventsApprovalsChecker.cs [hunk 2]
```diff
@@ -40,7 +76,7 @@ protected override async Task<string> GetNextApprover(string employeeId, IEnumer
 
                 finalApprover = parentDepartment?.ChiefId;
             }
-            else if (employeeDepartment.IsHeadDepartment && employeeDepartment.ChiefId == employeeId)
+            else if (employeeDepartment.IsHeadDepartment && isEmployeeChief)
             {
                 // No approvals required for Director General
```
Comment (santee) [inline server/Arcadia.Assistant.CSP/CspCalendarEventsApprovalsChecker.cs]: return null?

File: server/Arcadia.Assistant.CSP/CspCalendarEventsApprovalsChecker.cs [hunk 3]
```diff
@@ -40,7 +76,7 @@ protected override async Task<string> GetNextApprover(string employeeId, IEnumer
 
                 finalApprover = parentDepartment?.ChiefId;
             }
-            else if (employeeDepartment.IsHeadDepartment && employeeDepartment.ChiefId == employeeId)
+            else if (employeeDepartment.IsHeadDepartment && isEmployeeChief)
```
Comment (santee) [inline server/Arcadia.Assistant.CSP/CspCalendarEventsApprovalsChecker.cs]: do you also need to return null in here?

File: server/Arcadia.Assistant.CSP/CspCalendarEventsApprovalsChecker.cs (hunk 1 above)
Comment (santee) [reply server/Arcadia.Assistant.CSP/CspCalendarEventsApprovalsChecker.cs]: aslo you never check employeeMetadata for null

Comment (sponkrashin) [reply server/Arcadia.Assistant.CSP/CspCalendarEventsApprovalsChecker.cs]: Shit, I missed this. It was too late...

File: server/Arcadia.Assistant.CSP/CspCalendarEventsApprovalsChecker.cs (hunk 3 above)
Comment (sponkrashin) [reply server/Arcadia.Assistant.CSP/CspCalendarEventsApprovalsChecker.cs]: No, it is handled below.

File: server/Arcadia.Assistant.Calendar/CalendarEventsStorageBase.cs [hunk 4]
```diff
@@ -77,23 +86,162 @@ protected override void OnCommand(object message)
                     }
                     break;
 
+                case ApproveCalendarEvent msg:
+                    try
```
Comment (santee) [inline server/Arcadia.Assistant.Calendar/CalendarEventsStorageBase.cs]: 90 lines already. should we split it somehow? :)

Comment (sponkrashin) [reply server/Arcadia.Assistant.Calendar/CalendarEventsStorageBase.cs]: Probably, yes, but not now, ok? :)

File: server/Arcadia.Assistant.CSP/CspCalendarEventsApprovalsChecker.cs (hunk 1 above)
Comment (sponkrashin) [reply server/Arcadia.Assistant.CSP/CspCalendarEventsApprovalsChecker.cs]: Fixed both.

File: server/Arcadia.Assistant.Server.Console/akka.conf [hunk 5]
```diff
@@ -42,7 +42,7 @@ akka {
                 plugin-dispatcher = "akka.actor.default-dispatcher"
 
                 # connection string used for database access
-                connection-string = "<set it>"
+                connection-string = "Data Source=(local);Initial Catalog=ArcadiaAssistant;Persist Security Info=True;User ID=sa;Password=sa"
```
Comment (santee) [inline server/Arcadia.Assistant.Server.Console/akka.conf]: lol

File: server/Arcadia.Assistant.Calendar/Vacations/EmployeeVacationsActor.cs [hunk 6]
```diff
@@ -168,11 +133,32 @@ protected override void UpdateCalendarEvent(CalendarEvent oldEvent, CalendarEven
...
+            var @event = new UserGrantedCalendarEventApproval
+            {
+                EventId = message.Event.EventId,
+                TimeStamp = DateTimeOffset.Now,
+                ApproverId = message.ApproverId
+            };
+
+            this.Persist(@event, ev =>
+            {
+                this.OnUserGrantedVacationApproval(ev);
```
Comment (santee) [inline server/Arcadia.Assistant.Calendar/Vacations/EmployeeVacationsActor.cs]: implementations are the same, only difference is method called. Maybe better not override this but have that call overwritten instead?

Comment (sponkrashin) [reply server/Arcadia.Assistant.Calendar/Vacations/EmployeeVacationsActor.cs]: I thought about it. Made another way to not have Persist method call in base class. Easily can be done as you suggest.

Comment (sponkrashin) [reply server/Arcadia.Assistant.Calendar/Vacations/EmployeeVacationsActor.cs]: Done.

File: server/Arcadia.Assistant.Calendar/SickLeave/EmployeeSickLeaveActor.cs [hunk 7]
```diff
@@ -1,22 +1,26 @@
...
     using Arcadia.Assistant.Calendar.Abstractions;
+    using Arcadia.Assistant.Calendar.Abstractions.Messages;
     using Arcadia.Assistant.Calendar.SickLeave.Events;
     using Arcadia.Assistant.Notifications;
     using Arcadia.Assistant.Organization.Abstractions;
 
     public class EmployeeSickLeaveActor : CalendarEventsStorageBase
     {
         private EmployeeMetadata employee;
+        private readonly List<string> eventsToProcessApproversAfterRecover = new List<string>();
```
Comment (santee) [inline server/Arcadia.Assistant.Calendar/SickLeave/EmployeeSickLeaveActor.cs]: Hashset

Comment (sponkrashin) [reply server/Arcadia.Assistant.Calendar/SickLeave/EmployeeSickLeaveActor.cs]: Done.

**Pull request merged**